File: services/media-worker/pullvio_worker/visolix.py
from __future__ import annotations

import ipaddress
import socket
from dataclasses import dataclass
from pathlib import Path
from typing import Any
from urllib.parse import urljoin, urlsplit

import requests

from .domain import WorkerError


VISOLIX_API_BASE = "https://developers.visolix.com/api"
ALLOWED_VIDEO_FORMATS = {"360", "480", "720", "1080", "1440", "2160"}


@dataclass(frozen=True)
class VisolixSubmission:
    provider_job_id: str
    info: dict[str, Any]


@dataclass(frozen=True)
class VisolixProgress:
    progress: int
    download_url: str | None
    info: dict[str, Any]

    @property
    def completed(self) -> bool:
        return self.download_url is not None


class VisolixClient:
    def __init__(
        self,
        api_key: str,
        *,
        base_url: str = VISOLIX_API_BASE,
        session: requests.Session | None = None,
    ):
        if not api_key or len(api_key) > 500:
            raise RuntimeError("Visolix API secret is malformed")
        parsed = urlsplit(base_url)
        if parsed.scheme != "https" or not parsed.hostname:
            raise RuntimeError("Visolix API base URL must use HTTPS")
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._session = session or requests.Session()
        self._session.trust_env = False

    def submit(self, source_url: str, provider_format: str) -> VisolixSubmission:
        if provider_format not in ALLOWED_VIDEO_FORMATS:
            raise WorkerError("INVALID_OPTIONS", "Unsupported YouTube video quality")
        try:
            response = self._session.get(
                f"{self._base_url}/download",
                headers={
                    "X-API-KEY": self._api_key,
                    "X-PLATFORM": "youtube",
                    "URL": source_url,
                    "X-FORMAT": provider_format,
                    "User-Agent": "pullvio-media-worker/1",
                },
                timeout=(5, 30),
            )
        except requests.RequestException as exc:
            raise WorkerError(
                "PROVIDER_SUBMISSION_AMBIGUOUS",
                "YouTube provider submission outcome is unknown",
                retryable=False,
            ) from exc
        _raise_for_provider_status(response.status_code)
        payload = _json_object(response)
        provider_job_id = payload.get("id")
        if payload.get("success") not in {True, 1} or not isinstance(provider_job_id, str):
            raise WorkerError(
                "PROVIDER_RESPONSE_INVALID",
                "YouTube provider returned an invalid submission response",
                retryable=True,
            )
        return VisolixSubmission(
            provider_job_id=provider_job_id[:500],
            info=_safe_provider_info(payload.get("info")),
        )

    def progress(self, provider_job_id: str) -> VisolixProgress:
        if not provider_job_id or len(provider_job_id) > 500:
            raise WorkerError("PROVIDER_STATE_INVALID", "YouTube provider job ID is invalid")
        try:
            response = self._session.get(
                f"{self._base_url}/progress",
                params={"id": provider_job_id},
                headers={
                    "X-API-KEY": self._api_key,
                    "User-Agent": "pullvio-media-worker/1",
                },
                timeout=(5, 30),
            )
        except requests.RequestException as exc:
            raise WorkerError(
                "PROVIDER_UNAVAILABLE",
                "YouTube provider progress is temporarily unavailable",
                retryable=True,
            ) from exc
        _raise_for_provider_status(response.status_code)
        payload = _json_object(response)
        raw_progress = payload.get("progress", 0)
        if not isinstance(raw_progress, int) or not 0 <= raw_progress <= 1000:
            raise WorkerError(
                "PROVIDER_RESPONSE_INVALID",
                "YouTube provider returned invalid progress",
                retryable=True,
            )
        download_url = payload.get("download_url")
        if download_url is not None and not isinstance(download_url, str):
            raise WorkerError(
                "PROVIDER_RESPONSE_INVALID",
                "YouTube provider returned an invalid result URL",
                retryable=True,
            )
        if download_url is not None:
            _validate_public_https_url(download_url)
        return VisolixProgress(
            progress=raw_progress,
            download_url=download_url,
            info=_safe_provider_info(payload.get("info")),
        )


def provider_format_for(requested_quality: str) -> str:
    normalized = requested_quality.removesuffix("p")
    if normalized == "best":
        return "1080"
    if normalized not in ALLOWED_VIDEO_FORMATS:
        raise WorkerError("INVALID_OPTIONS", "Unsupported YouTube video quality")
    return normalized


def download_provider_result(
    url: str,
    destination: Path,
    max_output_bytes: int,
    *,
    session: requests.Session | None = None,
    max_redirects: int = 3,
) -> int:
    if max_output_bytes <= 0:
        raise ValueError("Maximum output size must be positive")
    client = session or requests.Session()
    client.trust_env = False
    current_url = url

    for redirect_count in range(max_redirects + 1):
        _validate_public_https_url(current_url)
        response = client.get(
            current_url,
            stream=True,
            allow_redirects=False,
            timeout=(10, 120),
            headers={"User-Agent": "pullvio-media-worker/1"},
        )
        if response.status_code in {301, 302, 303, 307, 308}:
            location = response.headers.get("Location")
            response.close()
            if not location or redirect_count >= max_redirects:
                raise WorkerError(
                    "PROVIDER_DOWNLOAD_FAILED",
                    "YouTube provider result redirected unexpectedly",
                    retryable=True,
                )
            current_url = urljoin(current_url, location)
            continue
        if response.status_code >= 400:
            status = response.status_code
            response.close()
            raise WorkerError(
                "PROVIDER_DOWNLOAD_FAILED",
                f"YouTube provider result returned HTTP {status}",
                retryable=status in {408, 425, 429} or status >= 500,
            )

        content_length = response.headers.get("Content-Length")
        if content_length and content_length.isdigit() and int(content_length) > max_output_bytes:
            response.close()
            raise WorkerError("OUTPUT_SIZE_LIMIT", "Output exceeds the supported size")

        total = 0
        try:
            with destination.open("wb") as output:
                for chunk in response.iter_content(chunk_size=1024 * 1024):
                    if not chunk:
                        continue
                    total += len(chunk)
                    if total > max_output_bytes:
                        raise WorkerError("OUTPUT_SIZE_LIMIT", "Output exceeds the supported size")
                    output.write(chunk)
        except Exception:
            destination.unlink(missing_ok=True)
            raise
        finally:
            response.close()
        if total <= 0:
            destination.unlink(missing_ok=True)
            raise WorkerError(
                "PROVIDER_DOWNLOAD_FAILED",
                "YouTube provider returned an empty file",
                retryable=True,
            )
        return total

    raise WorkerError(
        "PROVIDER_DOWNLOAD_FAILED",
        "YouTube provider result could not be downloaded",
        retryable=True,
    )


def _raise_for_provider_status(status_code: int):
    if status_code < 400:
        return
    if status_code == 402:
        raise WorkerError(
            "PROVIDER_BALANCE_EXHAUSTED",
            "YouTube downloads are temporarily unavailable",
            retryable=False,
        )
    if status_code in {401, 403}:
        raise WorkerError(
            "PROVIDER_AUTH_ERROR",
            "YouTube provider authentication failed",
            retryable=False,
        )
    raise WorkerError(
        "PROVIDER_UNAVAILABLE",
        f"YouTube provider returned HTTP {status_code}",
        retryable=status_code in {408, 425, 429} or status_code >= 500,
    )


def _json_object(response: requests.Response) -> dict[str, Any]:
    try:
        value = response.json()
    except (ValueError, requests.JSONDecodeError) as exc:
        raise WorkerError(
            "PROVIDER_RESPONSE_INVALID",
            "YouTube provider returned invalid JSON",
            retryable=True,
        ) from exc
    if not isinstance(value, dict):
        raise WorkerError(
            "PROVIDER_RESPONSE_INVALID",
            "YouTube provider returned invalid JSON",
            retryable=True,
        )
    return value


def _safe_provider_info(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict):
        return {}
    safe: dict[str, Any] = {}
    for key in ("title", "duration", "thumbnail"):
        item = value.get(key)
        if isinstance(item, (str, int, float, bool)) and len(str(item)) <= 2048:
            safe[key] = item
    return safe


def _validate_public_https_url(value: str):
    try:
        parsed = urlsplit(value)
        port = parsed.port
    except ValueError as exc:
        raise WorkerError("PROVIDER_RESULT_UNSAFE", "Provider result URL is invalid") from exc
    if (
        parsed.scheme != "https"
        or not parsed.hostname
        or parsed.username
        or parsed.password
        or port not in {None, 443}
        or len(value) > 4096
    ):
        raise WorkerError("PROVIDER_RESULT_UNSAFE", "Provider result URL is not allowed")
    try:
        addresses = socket.getaddrinfo(parsed.hostname, 443, type=socket.SOCK_STREAM)
    except socket.gaierror as exc:
        raise WorkerError(
            "PROVIDER_RESULT_UNAVAILABLE",
            "Provider result host could not be resolved",
            retryable=True,
        ) from exc
    if not addresses:
        raise WorkerError("PROVIDER_RESULT_UNSAFE", "Provider result host has no address")
    for address in addresses:
        ip = ipaddress.ip_address(address[4][0])
        if not ip.is_global:
            raise WorkerError("PROVIDER_RESULT_UNSAFE", "Provider result resolved to a private address")

File: lib/media/visolix.ts
import "server-only";
import type { Json } from "@/lib/database.types";
import type { SubmitJobInput } from "./contracts";

const DEFAULT_API_BASE = "https://developers.visolix.com/api";
const PROVIDER_TIMEOUT_MS = 30_000;
const YOUTUBE_VIDEO_FORMATS = new Set(["360", "480", "720", "1080", "1440", "2160"]);
const VISOLIX_PLATFORMS = new Set([
  "youtube",
  "instagram",
  "facebook",
  "tiktok",
  "snapchat",
  "okru",
]);

export class VisolixError extends Error {
  readonly code: string;
  readonly httpStatus: number | null;
  readonly outcomeKnown: boolean;
  readonly diagnostics: Json;

  constructor(
    code: string,
    message: string,
    options: {
      httpStatus?: number | null;
      outcomeKnown?: boolean;
      diagnostics?: Json;
    } = {},
  ) {
    super(message);
    this.name = "VisolixError";
    this.code = code;
    this.httpStatus = options.httpStatus ?? null;
    this.outcomeKnown = options.outcomeKnown ?? true;
    this.diagnostics = options.diagnostics ?? {};
  }
}

export type VisolixSubmission = {
  providerJobId: string | null;
  downloadUrl: string | null;
  info: Json;
};

export type VisolixProgress = {
  progress: number;
  downloadUrl: string | null;
  text: string | null;
  info: Json;
};

export function providerFormatFor(input: SubmitJobInput) {
  if (!VISOLIX_PLATFORMS.has(input.sourcePlatform)) {
    throw new VisolixError("UNSUPPORTED_SOURCE", "This media source is not supported.");
  }
  if (input.sourcePlatform !== "youtube") {
    if (input.mediaKind !== "video") {
      throw new VisolixError(
        "AUDIO_UNAVAILABLE",
        "Audio output is currently available for YouTube links only.",
      );
    }
    return "source";
  }
  if (input.mediaKind === "audio") return "mp3";
  const format = input.quality === "best" ? "1080" : input.quality.replace(/p$/, "");
  if (!YOUTUBE_VIDEO_FORMATS.has(format)) {
    throw new VisolixError("INVALID_OPTIONS", "This YouTube quality is not supported.");
  }
  return format;
}

export class VisolixClient {
  private readonly apiKey: string;
  private readonly apiBase: string;

  constructor(apiKey = process.env.VISOLIX_API_KEY, apiBase = process.env.VISOLIX_API_BASE_URL) {
    if (!apiKey || apiKey.length > 500) {
      throw new VisolixError("PROVIDER_NOT_CONFIGURED", "The media provider is not configured.");
    }
    this.apiKey = apiKey;
    this.apiBase = validateApiBase(apiBase || DEFAULT_API_BASE);
  }

  async submit(input: SubmitJobInput): Promise<VisolixSubmission> {
    const providerFormat = providerFormatFor(input);
    const headers: Record<string, string> = {
      "X-API-KEY": this.apiKey,
      "X-PLATFORM": input.sourcePlatform,
      URL: input.sourceUrl,
      "User-Agent": "pullvio-vercel/1",
    };
    if (input.sourcePlatform === "youtube") headers["X-FORMAT"] = providerFormat;

    const response = await this.request(`${this.apiBase}/download`, { headers }, true);
    const payload = await jsonObject(response);
    const success = payload.success;
    const providerJobId = normalizeProviderJobId(payload.id);
    const preferredMediaKeys = input.sourcePlatform === "youtube"
      ? input.mediaKind === "audio"
        ? ["mp3"]
        : [providerFormat, `${providerFormat}p`, "hd", "sd"]
      : ["hd", "sd"];
    const synchronousUrl = extractSynchronousUrl(payload.media, preferredMediaKeys);
    if (synchronousUrl) {
      return {
        providerJobId: null,
        downloadUrl: synchronousUrl,
        info: safeInfo(payload),
      };
    }
    if (success === false || success === 0 || success === "0" || success === "false") {
      throw new VisolixError("PROVIDER_REJECTED", "The provider rejected this media.", {
        httpStatus: response.status,
      });
    }
    if (
      !isSuccessfulFlag(success)
      || !providerJobId
    ) {
      throw new VisolixError("PROVIDER_RESPONSE_INVALID", "The provider returned an invalid response.", {
        httpStatus: response.status,
        outcomeKnown: false,
        diagnostics: responseShape(payload),
      });
    }
    return {
      providerJobId,
      downloadUrl: null,
      info: safeInfo(payload.info),
    };
  }

  async progress(providerJobId: string): Promise<VisolixProgress> {
    if (!providerJobId || providerJobId.length > 500) {
      throw new VisolixError("PROVIDER_STATE_INVALID", "The provider job ID is invalid.");
    }
    const endpoint = new URL(`${this.apiBase}/progress`);
    endpoint.searchParams.set("id", providerJobId);
    const response = await this.request(endpoint.toString(), {
      headers: {
        "X-API-KEY": this.apiKey,
        "User-Agent": "pullvio-vercel/1",
      },
    });
    const payload = await jsonObject(response);
    const progress = payload.progress ?? 0;
    const success = payload.success;
    if (!Number.isInteger(progress) || Number(progress) < 0 || Number(progress) > 1000) {
      throw new VisolixError("PROVIDER_RESPONSE_INVALID", "The provider returned invalid progress.");
    }
    if (![0, 1, false, true].includes(success as boolean | number)) {
      throw new VisolixError("PROVIDER_RESPONSE_INVALID", "The provider returned an invalid status.");
    }

    const rawUrl = payload.download_url === "" ? null : payload.download_url;
    if (rawUrl !== null && rawUrl !== undefined && typeof rawUrl !== "string") {
      throw new VisolixError("PROVIDER_RESPONSE_INVALID", "The provider returned an invalid download URL.");
    }
    const downloadUrl = rawUrl ? validateResultUrl(rawUrl) : null;
    if ((success === 1 || success === true) && !downloadUrl) {
      throw new VisolixError("PROVIDER_RESPONSE_INVALID", "The completed provider job has no result URL.");
    }
    return {
      progress: Number(progress),
      downloadUrl,
      text: typeof payload.text === "string" ? payload.text.slice(0, 300) : null,
      info: safeInfo(payload.info),
    };
  }

  private async request(url: string, init: RequestInit, submission = false) {
    let response: Response;
    try {
      response = await fetch(url, {
        ...init,
        cache: "no-store",
        signal: AbortSignal.timeout(PROVIDER_TIMEOUT_MS),
      });
    } catch {
      throw new VisolixError(
        submission ? "PROVIDER_SUBMISSION_AMBIGUOUS" : "PROVIDER_UNAVAILABLE",
        submission
          ? "The provider submission outcome is unknown."
          : "The provider is temporarily unavailable.",
        { outcomeKnown: !submission },
      );
    }

    if (response.ok) return response;
    if (response.status === 402) {
      throw new VisolixError(
        "PROVIDER_BALANCE_EXHAUSTED",
        "Media downloads are temporarily unavailable.",
        { httpStatus: 402 },
      );
    }
    if (response.status === 400) {
      throw new VisolixError("PROVIDER_REJECTED", "The provider rejected this media.", {
        httpStatus: 400,
      });
    }
    if (response.status === 401 || response.status === 403) {
      throw new VisolixError("PROVIDER_AUTH_FAILED", "The provider authentication failed.", {
        httpStatus: response.status,
      });
    }
    throw new VisolixError("PROVIDER_UNAVAILABLE", "The provider is temporarily unavailable.", {
      httpStatus: response.status,
      outcomeKnown: !submission,
    });
  }
}

function isSuccessfulFlag(value: unknown) {
  return value === true || value === 1 || value === "1" || value === "true";
}

function normalizeProviderJobId(value: unknown) {
  if (typeof value === "string") {
    const normalized = value.trim();
    return normalized.length >= 1 && normalized.length <= 500 ? normalized : null;
  }
  if (typeof value === "number" && Number.isSafeInteger(value) && value >= 0) {
    return String(value);
  }
  return null;
}

function extractSynchronousUrl(value: unknown, preferredKeys: string[] = []): string | null {
  if (typeof value === "string" && value) return validateResultUrl(value);
  if (Array.isArray(value)) {
    for (const item of value.slice(0, 20)) {
      const candidate = extractSynchronousUrl(item, preferredKeys);
      if (candidate) return candidate;
    }
    return null;
  }
  if (!value || typeof value !== "object") return null;
  const object = value as Record<string, unknown>;
  for (const key of [
    ...preferredKeys,
    "url",
    "download_url",
    "downloadUrl",
    "video_url",
    "videoUrl",
    "file",
    "link",
    "media",
  ]) {
    const candidate = object[key];
    const resolved = extractSynchronousUrl(candidate);
    if (resolved) return resolved;
  }
  return null;
}

function responseShape(payload: Record<string, unknown>): Json {
  const shape: Record<string, Json> = {
    keys: Object.keys(payload).slice(0, 20),
    successType: typeof payload.success,
    idType: Array.isArray(payload.id) ? "array" : typeof payload.id,
    statusType: typeof payload.status,
    statusValue: ["string", "number", "boolean"].includes(typeof payload.status)
      ? String(payload.status).slice(0, 50)
      : null,
  };
  for (const key of ["data", "result", "media"]) {
    const value = payload[key];
    if (value && typeof value === "object" && !Array.isArray(value)) {
      shape[`${key}Keys`] = Object.keys(value as Record<string, unknown>).slice(0, 20);
    } else if (Array.isArray(value)) {
      shape[`${key}Type`] = "array";
      const first = value[0];
      if (first && typeof first === "object" && !Array.isArray(first)) {
        shape[`${key}ItemKeys`] = Object.keys(first as Record<string, unknown>).slice(0, 20);
      } else {
        shape[`${key}ItemType`] = typeof first;
      }
    } else {
      shape[`${key}Type`] = Array.isArray(value) ? "array" : typeof value;
    }
  }
  return shape;
}

function validateApiBase(value: string) {
  const parsed = new URL(value);
  if (
    parsed.protocol !== "https:"
    || parsed.hostname !== "developers.visolix.com"
    || parsed.username
    || parsed.password
  ) {
    throw new VisolixError("PROVIDER_NOT_CONFIGURED", "The provider API base URL is invalid.");
  }
  return parsed.toString().replace(/\/$/, "");
}

function validateResultUrl(value: string) {
  if (value.length > 4096) {
    throw new VisolixError("PROVIDER_RESPONSE_INVALID", "The provider result URL is too long.");
  }
  const parsed = new URL(value);
  if (parsed.protocol !== "https:" || !parsed.hostname || isPrivateHostname(parsed.hostname)) {
    throw new VisolixError("PROVIDER_RESPONSE_INVALID", "The provider result URL is not public HTTPS.");
  }
  return parsed.toString();
}

function isPrivateHostname(hostname: string) {
  const normalized = hostname.toLowerCase().replace(/\.$/, "");
  if (
    normalized === "localhost"
    || normalized === "::1"
    || normalized === "[::1]"
    || normalized.endsWith(".local")
  ) return true;
  if (
    /^0\./.test(normalized)
    || /^127\./.test(normalized)
    || /^10\./.test(normalized)
    || /^169\.254\./.test(normalized)
    || /^192\.168\./.test(normalized)
  ) return true;
  const match = normalized.match(/^172\.(\d{1,3})\./);
  return Boolean(match && Number(match[1]) >= 16 && Number(match[1]) <= 31);
}

async function jsonObject(response: Response): Promise<Record<string, unknown>> {
  let payload: unknown;
  try {
    payload = await response.json();
  } catch {
    throw new VisolixError("PROVIDER_RESPONSE_INVALID", "The provider returned invalid JSON.", {
      httpStatus: response.status,
      outcomeKnown: false,
    });
  }
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    throw new VisolixError("PROVIDER_RESPONSE_INVALID", "The provider returned invalid JSON.", {
      httpStatus: response.status,
      outcomeKnown: false,
    });
  }
  return payload as Record<string, unknown>;
}

function safeInfo(value: unknown): Json {
  if (!value || typeof value !== "object" || Array.isArray(value)) return {};
  const input = value as Record<string, unknown>;
  const output: Record<string, Json> = {};
  for (const key of ["title", "thumbnail", "duration", "author", "platform"]) {
    const item = input[key];
    if (typeof item === "string") output[key] = item.slice(0, key === "thumbnail" ? 2048 : 500);
    else if (typeof item === "number" && Number.isFinite(item)) output[key] = item;
  }
  if (!output.thumbnail && typeof input.thumb === "string") {
    output.thumbnail = input.thumb.slice(0, 2048);
  }
  if (!output.platform && typeof input.service === "string") {
    output.platform = input.service.slice(0, 500);
  }
  return output;
}
